libuavcan/src/transport/uc_can_acceptance_filter_configurator.cpp

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (C) 2014 Pavel Kirienko <pavel.kirienko@gmail.com>,
+ *                    Ilia  Sheremet <illia.sheremet@gmail.com>
+ */
+
+#include <uavcan/transport/can_acceptance_filter_configurator.hpp>
+#include <cassert>
+
+namespace uavcan
+{
+const unsigned CanAcceptanceFilterConfigurator::DefaultFilterMsgMask;
+const unsigned CanAcceptanceFilterConfigurator::DefaultFilterServiceRequestID;
+const unsigned CanAcceptanceFilterConfigurator::DefaultFilterServiceRequestMask;
+const unsigned CanAcceptanceFilterConfigurator::ServiceRespFrameID;
+const unsigned CanAcceptanceFilterConfigurator::ServiceRespFrameMask;
+
+int16_t CanAcceptanceFilterConfigurator::loadInputConfiguration()
+{
+    multiset_configs_.clear();
+
+    CanFilterConfig service_resp_cfg;
+    service_resp_cfg.id = ServiceRespFrameID;
+    service_resp_cfg.mask = ServiceRespFrameMask;
+	if (multiset_configs_.emplace(service_resp_cfg) == NULL)
+    {
+        return -ErrMemory;
+    }
+
+    const TransferListenerBase* p = node_.getDispatcher().getListOfMessageListeners().get();
+    while (p)
+    {
+        CanFilterConfig cfg;
+        cfg.id = static_cast<uint32_t>(p->getDataTypeDescriptor().getID().get()) << 16;
+        cfg.id |= static_cast<uint32_t>(p->getDataTypeDescriptor().getKind()) << 8;
+        cfg.mask = DefaultFilterMsgMask;
+        if (multiset_configs_.emplace(cfg) == NULL)
+            {
+                return -ErrMemory;
+            }
+        p = p->getNextListNode();
+    }
+
+    const TransferListenerBase* p1 = node_.getDispatcher().getListOfServiceRequestListeners().get();
+    while (p1)
+    {
+        CanFilterConfig cfg;
+        cfg.id = DefaultFilterServiceRequestID;
+        cfg.id |= static_cast<uint32_t>(p1->getDataTypeDescriptor().getID().get()) << 17;
+        cfg.mask = DefaultFilterServiceRequestMask;
+        if (multiset_configs_.emplace(cfg) == NULL)
+            {
+                return -ErrMemory;
+            }
+        p1 = p1->getNextListNode();
+    }
+
+    if (multiset_configs_.getSize() == 0)
+    {
+        return -ErrLogic;
+    }
+
+#if UAVCAN_DEBUG
+    for (uint16_t i = 0; i < multiset_configs_.getSize(); i++)
+    {
+        UAVCAN_TRACE("CanAcceptanceFilterConfigurator::loadInputConfiguration()", "cfg.ID [%u] = %d", i,
+                     multiset_configs_.getByIndex(i)->id);
+        UAVCAN_TRACE("CanAcceptanceFilterConfigurator::loadInputConfiguration()", "cfg.MK [%u] = %d", i,
+                     multiset_configs_.getByIndex(i)->mask);
+    }
+#endif
+
+    return 0;
+}
+
+int16_t CanAcceptanceFilterConfigurator::computeConfiguration()
+{
+    uint8_t i_rank = 0, j_rank = 0;
+    const uint16_t acceptance_filters_number = getNumFilters();
+
+    while (acceptance_filters_number < multiset_configs_.getSize())
+    {
+        UAVCAN_ASSERT(multiset_configs_.getSize() != 0);
```

Comment:
(static analyzer hat on): This check contradicts the loop termination condition. Probably you wanted to put it before the loop?